UI merge

RedRafe opened the pull request:
## Brief description of the changes:

- [x] Added formatted tooltips to all top UI buttons
- [x] Changed `gui.top` and `gui.left` to use `mod_gui`'s flows instead
- [x] Changed most GUIs to use shared `utils.gui` and `utils.utils.gui_style` instead
- [x] Removed `on_player_created` handler for GUI restyle
- [x] Changed *"Fish menu"* name from "Comfy Panel" to "Menu & Settings"
- [x] Changed *"Fish menu"* icon from raw-fish to Factorio gear logo
- [x] Changed *"Team Manager"* button from text to "Red Flag" icon
- [x] Changed *"Research Info"* button behavior to toggle on/off the window instead of bringing it forward on multiple clicks
- [x] Changed *"Research Info"* sprite from `space-science-pack` to `lab`
- [x] Changed *"Map info"* icon from question mark to book (`custom_tag_icon`)
- [x] Changed *"Poll"* windows
- [x] Fixed incorrect Poll element reference
- [x] Changed *"Suspend"* UI frame
- [x] Changed *"Reroll"* UI frame
- [x] Changed *main left gui* style & contents (refactored whole UI and buttons)
  - [x] Changed text info to buttons (player count, evolution, threat)
  - [x] Added additional info for threat and evolution in buttons' tooltip
  - [x] Removed per-player data of show/hide teams player list
  - [x] Added captain label for cpt mode in player list
  - [x] Moved join/resume buttons from team section to a separate subsection of the gui
  - [x] Changed spectate/resume buttons
  - [x] Changed disabled foods from disabled buttons to a new info-alert button
  - [x] Added *"Research Info"* button to *main left gui*
  - [x] Added *"Team Statistics"* button to *main left gui*
  - [x] Added *"Shortcuts"* button to *main left gui*
- [x] Changed Captain game class 
  - [x] Added `gui.left` and `gui.screen` menu for cpt games
  - [x] Added auto-render for cpt pages (toggle/draw/update) 
  - [x] Added new info panel for cpt games  
  - [x] Changed & merged CPT buttons into new *"Tournament"* button
  - [x] Changed cpt join window 
  - [x] Changed *"Team Organization* class
  - [x] Changed picking window for captains to maintain screen location
  - [x] Changed main frames to closable frames 
- [x] Changed & merged MVPs/winning team gui
- [x] Added top UI theme selector in configs
- [x] ~~Added floating shortcuts and per-player configs for them~~
- [x] Added vote counts to reroll window/suspend windows
- [x] Changed some texts to lacalised strings
- [x] Removed unused locale files  

## Tested Changes:
- [x] I've tested the changes locally in SP
- [x] I've tested up to 9th draft with ppl in MP
- [x] Need testing after review

clifffrey commented on maps/biter_battles_v2/team_stats_compare.lua:
```diff
@@ -257,6 +257,42 @@ function TeamStatsCompare.game_over()
+        Sounds.notify_player(player, 'utility/cannot_build')
+        return
+    end
+    safe_wrap_with_player_print(player, TeamStatsCompare.show_stats, player, stats)
```
it would be better to move this up a level, i.e. `safe_wrap_with_player_print(player, TeamStatsCompare.toggle_team_stats, player, stats)` and instead just directly call `show_stats` here. That way, if there are lua errors in the code in this function, it will not crash the game.

(this would just be an improvement that is somewhat outside the scope of your change, but related to this refactor)

clifffrey commented on modules/simple_tags.lua:
```diff
@@ -53,7 +54,7 @@ local function draw_screen_gui(player)
 		direction = "vertical",
 	})	
-	frame.location = {x = get_x_offset(player) * player.display_scale - 2, y = 39 * player.display_scale}
+	frame.location = {x = get_x_offset(player) * player.display_scale - 2, y = --[[39]] 54 * player.display_scale}
```
remove comment?

clifffrey commented on utils/color_presets.lua:
```diff
@@ -1,157 +1,309 @@
+    captain_versus_float = {r = 0.96, g = 0.85, b = 0.13, a = 1 },
 }
+--[[{
+    maroon = { 128, 0, 0 },
```
I'm not sure why we are adding this?

RedRafe replied:
My bad, that was because my editor highlights the rgb values of that other format and so it was easier for me to find color, but the I forgot to remove it. dunno if you have another extension that highlights the explicit version (maybe I should check if mine supports other formats, I didn't check before tbh)

![Screenshot from 2024-08-01 12-29-14](https://github.com/user-attachments/assets/900ea9f8-ba6a-47bb-bedf-2fe3fc157ecd)

clifffrey commented on utils/gui.lua:
```diff
@@ -115,6 +122,135 @@ function Gui.clear(element)
+end
+
+---@param player LuaPlayer
+---@param frame LuaGuiElement|table
```
I think that this would be more precise (and maybe give better help text/completions) if you had this as `LuaGuiElement|LuaGuiElement.add_param`).

This comment applies a few other places in the file as well.

clifffrey commented on utils/gui.lua:
```diff
@@ -115,6 +122,135 @@ function Gui.clear(element)
+
+---@param player LuaPlayer
+---@param element_name string
+---@return LuaGuiElement|nil
```
I feel like this is normally written as `---@return LuaGuiElement?`

clifffrey commented on utils/gui.lua:
```diff
@@ -115,6 +122,135 @@ function Gui.clear(element)
+        return element
+    end
+
+    for _, e in pairs(element.children) do
```
it is a bit surprising to me that you have both this block, and the block on lines 233-235.  Can you perhaps remove one or the other?

clifffrey commented on utils/gui.lua:
```diff
@@ -115,6 +122,135 @@ function Gui.clear(element)
+    return pusher
+end
+
+function Gui.get_child_recursively(element, child_name)
```
this function is never called I think?  remove?

clifffrey commented on utils/player.lua:
```diff
@@ -1,11 +1,24 @@
 local Public = {}
 
+local function cpt_get_player(playerName)
```
I'd slightly prefer that this be simplified to just essentially be
`return game.get_player(playerName) or { ... something that defines whatever functions are needed, I think just color? }` and then rename the function to something like `get_player_color`

Basically, I don't think that we need to expose the `captain_mode` stuff at all here - we can just accept any player name (and just default to a boring color?)

clifffrey reviewed (commented):
I am relatively unhappy about the large amount of reformatting done in this change. It makes it much harder to review.

If you are ever inclined to do lots of reformatting, please please reformat a file/function/etc in a change all by itself, and submit that change.
When reviewing a change, I rely on red/green lines to tell me what to look at. The formatting changes really get in the way of that.  (this is made slightly better by having the diff-program ignore whitespace, but it is still much simpler for them to be in changes by themselves)

Your whitespace changes also are using 2-space-indent, which is essentially otherwise unused in our files. I have a preference towards using either tabs, 4 spaces, or preparing a change that reformats our entire repository to 2 spaces (or something else).  Essentially almost all of the code (at least in the files you modified) is either "tabs" or "4 spaces" today - I don't want to add a 3rd configuration.  That said, please do not make any other large whitespace changes until we are close to the end of the code review cycle.

Have you verified that `lua ./tests/test-feeding.lua` still works?

With the immense size of this change, I worry that I may have missed something. I want to point out that many of the messages that are logged to discord, we are very sensitive to exactly what text is in them, because the ravenpatrol website has code that parses messages out of discord. So, I'd love some reassurance that you do not believe that you changed any discord log messages.

Thank you for removing a bunch of files that are clearly not needed.

I have not reviewed gui.lua, or the captains files yet. I will do so later. The change is big.

RedRafe commented:
1. **Logs** - I believe I intentionally didn't touch any of the messages logged through server/discord knowing they are sensitive text, but I will take my time to check again myself that all has remained the same

2. **Coding Conventions** - I see a lot of comments regarding coding style, and I know everyone has theirs preferences, but some files (captains.lua in particularly) had become a big clusterfuck of everyone-put-their-hands-on-it and there was total anarchy and working on it was really impossible... It's a 2k lines file, I had to touch most of it, so I took the time and carefully rearranged it (alphabetical imports, local scopes, classes, code conventions...). At least now 95% of the time all functions are `snake_case`, classes have the `PascalCase` format, and variables are in `camelCase` as of the rest of the repo, so it's somewhat readable. I know you prefer table['field'] when accessing tables, but even with advanced IDEs, readability of this codebase was severely impaired due to the amount of string concatenations and string texts thrown around. here's a couple examples:

![create_cpt](https://github.com/user-attachments/assets/7e695b02-b0c5-43a2-a50b-628f5c784c09)


Texts displayed to the user immediately jump out in the code instead of getting mixed with property accessing. It is also very important when accessing a class method via text because when accessing a method via concatenation i.e. buttons do `Public['draw_'..variable](player, event)` to be more dynamic, which is really important to catch in the code but is instead masked by a lot of `global['captain_variables']['special_variable]` that could simply be `global.captain_variables.special_variable` since they are clear properties defined when creating such tables.
Also there's a lot of `global.tableOfStuff['subName']['subName2]` which is a mixed behavior, and lot of GUI stuff is something like 'frame.flow.scroll.table.button...` which changing all these codes to be 'frame['flow']['scroll']['table`]['button'] really would make it so much harder to read.

3. **Localised Strings**: I don't want to go full-blown on localised strings, but they are quite handy in few places. The PROs for me are:
  - when using the same chunk of text in multiple places, you guarantee consistency across all instances of that text. Also improves maintanability, as you only need to change the text in one place instead of running around through the code and searching all the places where it may or may not appear
  - improves readability with long chunk of text, especially when concatenation and formatting is required. All the formatting can be moved to the locale file, while in the code only remains the bone of the logic, for example:
  
  `game.print(playerOfCommand.name .. " has decided that " .. victim.name .. " will be the new referee instead of " .. global.special_games_variables["captain_mode"]["refereeName"],Color.cyan)`

becomes simply:
  
  `game.print({'captain.replace_referee_announcement', playerOfCommand.name, victim.name, special.refereeName})`
  
  - it's true we don't have any current plans of supporting translations, but it is also true we didn't have any text to be translated before. The only localised text we had was the intro text, and that got translated into multiple languages already. I don't mean to translate everything to other languages, but if ppl will feel the need to do so, there will be an infrastructure to support so

![command](https://github.com/user-attachments/assets/256cb869-a486-4381-99f1-67d96bf932f1)

4. **Test Feed** - I didn't test it and I see `master` nor my branch have this `lunatest` file so maybe it's something we can have a look at?

clifffrey commented:
> 1. **Logs** - I believe I intentionally didn't touch any of the messages logged through server/discord knowing they are sensitive text, but I will take my time to check again myself that all has remained the same

Great.

> 2. **Coding Conventions** - I see a lot of comments regarding coding style, and I know everyone has theirs preferences, but some files (captains.lua in particularly) had become a big clusterfuck of everyone-put-their-hands-on-it and there was total anarchy and working on it was really impossible... It's a 2k lines file, I had to touch most of it, so I took the time and carefully rearranged it (alphabetical imports, local scopes, classes, code conventions...). At least now 95% of the time all functions are `snake_case`, classes have the `PascalCase` format, and variables are in `camelCase` as of the rest of the repo, so it's somewhat readable. I know you prefer table['field'] when accessing tables, but even with advanced IDEs, readability of this codebase was severely impaired due to the amount of string concatenations and string texts thrown around. here's a couple examples:
> 
> Texts displayed to the user immediately jump out in the code instead of getting mixed with property accessing. It is also very important when accessing a class method via text because when accessing a method via concatenation i.e. buttons do `Public['draw_'..variable](player, event)` to be more dynamic, which is really important to catch in the code but is instead masked by a lot of `global['captain_variables']['special_variable]` that could simply be `global.captain_variables.special_variable` since they are clear properties defined when creating such tables. Also there's a lot of `global.tableOfStuff['subName']['subName2']` which is a mixed behavior, and lot of GUI stuff is something like `frame.flow.scroll.table.button...` which changing all these codes to be `frame['flow']['scroll']['table']['button']` really would make it so much harder to read.

For my comments here, I completely agree with you about cases like `global['captain_variables']['special_variable']` being confusing/bad/unclear. For me it is when you have code like the following:
```
  myDataStructure[player.force.name] = 1
  ...
  myDataStructure.north = 2
```
I would rather that we consistently treat `myDataStructure` as if it is a `table<string, number>`, and thus always use the `['north']` notation. I really think about it in terms of types. If something is a struct that contains two fields, `.playtime` and `.buildings_placed`, then I strongly-prefer that we used the dot-notation. If something is a table that maps force-name to something else, I prefer to use the `['north']` notation.

Thus, it is only in the few places where I put comments that I lean towards changing it back. 99% of the places where you changed it I agree with and like the change.

I appreciate that you cleaned up many of the names in these files.

> 
> 3. **Localised Strings**: I don't want to go full-gas on localised strings either, but they are quite handy in few places. The PROs for me are:
> 
> * when using the same chunk of text in multiple places, you guarantee consistency across all instances of that text. Also improves maintanability, as you only need to change the text in one place instead of running around through the code and searching all the places where it may or may not appear
> * improves readability with long chunk of text, especially when concatenation and formatting is required. All the formatting can be moved to the locale file, while in the code only remains the bone of the logic, for example:
> 
> `game.print(playerOfCommand.name .. " has decided that " .. victim.name .. " will be the new referee instead of " .. global.special_games_variables["captain_mode"]["refereeName"],Color.cyan)`
> 
> becomes simply:
> 
> `game.print({'captain.replace_referee_announcement', playerOfCommand.name, victim.name, special.refereeName})`

My personal preference would instead be for it to use `string.format` in an intelligent way, which I then find even more readable, as with the "simple" print example you give above, I can't know anything about the argument order, it is very hard for me to verify if the referee and victim names are backwards or not.
If one print is used many times, then a simple function in the file can avoid the redundancy.  By using the locale infrastructure, it puts the strings in a very far away place, that my editor cannot quickly jump to the correct location for.  Thus, I would prefer if we reverted those parts of this change. That said, I will not insist on it. (I'm still very excited about the new UI, and I am willing to pay the price of locale use to get it if necessary).

> 
> * it's true we don't have any current plans of supporting translations, but it is also true we didn't have any text to be translated before. The only localised text we had was the intro text, and that got translated into multiple languages already. I don't mean to translate everything to other languages, but if ppl will feel the need to do so, there will be an infrastructure to support so
> 
> ![command](https://private-user-images.githubusercontent.com/93430988/353347945-256cb869-a486-4381-99f1-67d96bf932f1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjIzODQ5NzYsIm5iZiI6MTcyMjM4NDY3NiwicGF0aCI6Ii85MzQzMDk4OC8zNTMzNDc5NDUtMjU2Y2I4NjktYTQ4Ni00MzgxLTk5ZjEtNjdkOTZiZjkzMmYxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MzElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzMxVDAwMTExNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTk3NWVmNTZiMDNjZWQ0YTI2OTEyODVmNGViY2VlN2IwZTYxMGRkYWNhNDUwMTJmMGZhZTMzNzFjYjFhMDZmNDUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.jWB5BZQJQq-mxlKKeDe_Z-X6VhZlp2czUYBkg6L_rug)
> 
> 4. **Test Feed** - I didn't test it and I see `master` nor my branch have this `lunatest` file so maybe it's something we can have a look at?

If you run something like `luarocks install lunatest` you should then be able to run the test.

clifffrey commented on maps/biter_battles_v2/gui.lua:
```diff
@@ -1,47 +1,68 @@
+	blue = function(str) return '[color=blue]'..str..'[/color]' end,
+}
+
+local TEST_1 = '[color=acid]cliff_build[/color] [color=red]Carl3[/color] [color=white]Blitzi__[/color] [color=pink]califerbopbp[/color] [color=blue]Jimmy50[/color] [color=red]Ncling[/color] [color=acid]thesoldier_57[/color] [color=gray]sticklord[/color] [color=red]TheCakeISALie[/color] [color=yellow]Zerosun[/color] [color=green]kingofgugu[/color] [color=cyan]PaveV[/color] [color=cyan]cojito[/colo ...
```
I tend to avoid using real-player-names unless I have a strong reason to...  But maybe that is silly. Anyhow, you should think about replacing these with generic names.

clifffrey reviewed (commented):
Ok, I looked over the last 3 files. Definitely not a super close review as the changes are so big, but I'm not worried.

to brainstorm some things that we should manually test when we have multiple people on a server, I made a list.  Probably these are all tested already, it is fine if we don't do them, but I wanted to write down a list as I thought of them.

1. suspend-vote
2. group-picking in captains games
3. /replaceCaptainNorth
4. /replaceReferee
5. Use task groups some

RedRafe commented on maps/biter_battles_v2/game_over.lua:
```diff
@@ -1,625 +1,664 @@
+    for _, player in pairs(game.connected_players) do
+      local frame = Gui.get_top_element(player, 'reroll_frame')
+      if frame and frame.valid then
+        frame.flow.reroll_table.children[1].caption = {'gui.reroll_caption', global.reroll_time_left}
```
I noticed as well but this is how it was, so, heh

RedRafe commented on maps/biter_battles_v2/research_info.lua:
```diff
@@ -484,12 +487,19 @@ end
+    if frame and frame.valid then
+        --player.gui.screen["research_info_frame"].bring_to_front()
+        --player.gui.screen["research_info_frame"].force_auto_center()
+        if player.opened == frame then
```
All GUI toggles follow that same structure:

```
if frame then
    destroy()
else
    create()
end
```
but, every frame may have different triggers or different ways to get the main frame, so I'm not sure moving it to a general function is actually less code. Your call tho

clifffrey replied:
I was specifically talking about the logic that is interacting with `player.opened` which feels to me like it belongs in closable_frame.lua.

So, it is not so much about reducing the total amount of code, as it is about having the responsibility for managing player.opened in one place.

RedRafe replied:
@clifffrey I actually took the time to check, and when deregistering a Gui, destroying it already sets `player.open = nil` by the engine if that was the currently opened gui. So I've removed that part completely as it is completely redundant

RedRafe commented on comfy_panel/special_games/captain.lua:
This will be 

```
commands.add_command(...
  if game.is_multiplayer() then game.print(...) return end
  ...
 )
 ```
 
 Because game is not available on bootstrap (so the command is always added and if someone types it, it gets the alert message)

RedRafe commented:
Ok, fixed most of the issues, remaining topics:
- localised strings: I'm not very inclined to revert those to string formats, some are really really long and it is very hard to read with all the [font=] and [color=] formatting. Maybe some of the shorterst ones
- add or not add a function in `ClosableFrame` class to handle toggling frames

clifffrey commented:
> Ok, fixed most of the issues, remaining topics:
> 
> * localised strings: I'm not very inclined to revert those to string formats, some are really really long and it is very hard to read and code them in line with all the [font=] and [color=] formatting. Maybe some of the shortest ones
> * add or not add a function in `ClosableFrame` class to handle toggling frames

ok, on both of these topics I think that I've made my opinion clear, but I am also being clear that I'm not going to block merging. So if you are happy with this PR, then that is good enough for me.

RedRafe commented:
Removed a bunch of redundant logic/comments with the 13th draft

clifffrey reviewed (approved):
I'll vaguely plan on rolling this out at the end of the weekend or on Monday/Tuesday, but this is assuming that we can do another multiplayer test of the code before then.

RedRafe commented:
- removed switch for Sound FX (only used by 1 out of 633 last check) - establishing default as always enable for everyone. It is still possible to mute or turn up/down each individual sound channel from the Factorio Menu --> Settings --> Sounds
- removed shortcuts switch from config panel and from left gui, now only availale as button on the left gui instead
